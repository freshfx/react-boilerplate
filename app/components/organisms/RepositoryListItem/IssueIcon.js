import styled from '@emotion/styled'

import NormalIssueIcon from 'components/atoms/IssueIcon'

const IssueIcon = styled(NormalIssueIcon)`
  fill: #ccc;
  margin-right: 0.25em;
`

export default IssueIcon
