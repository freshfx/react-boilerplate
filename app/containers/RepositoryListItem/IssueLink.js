import styled from '@emotion/styled'
import NormalA from 'components/Anchor'

const IssueLink = styled(NormalA)`
  height: 100%;
  color: black;
  display: flex;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
`

export default IssueLink
