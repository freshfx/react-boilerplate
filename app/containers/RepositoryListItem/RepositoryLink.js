import styled from '@emotion/styled'
import NormalA from 'components/Anchor'

const RepositoryLink = styled(NormalA)`
  height: 100%;
  color: black;
  display: flex;
  align-items: center;
  width: 100%;
`

export default RepositoryLink
