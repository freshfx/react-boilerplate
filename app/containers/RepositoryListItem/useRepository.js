import {useSelector} from 'react-redux'

import {selectors} from 'modules/repository/entity'
import {selectors as homePageSelectors} from 'modules/pages/home'
import useMemoizedSelector from 'utils/useMemoizedSelector'

const useRepositoryState = id => ({
  fullName: useMemoizedSelector(selectors.makeSelectFullName, {id}),
  name: useMemoizedSelector(selectors.makeSelectName, {id}),
  openIssuesCount: useMemoizedSelector(selectors.makeSelectOpenIssuesCount, {id}),
  ownerUsername: useMemoizedSelector(selectors.makeSelectOwnerUsername, {id}),
  url: useMemoizedSelector(selectors.makeSelectUrl, {id}),
  username: useSelector(homePageSelectors.selectUsername)
})

const useRepository = id => {
  const state = useRepositoryState(id)
  return [state]
}

export default useRepository
