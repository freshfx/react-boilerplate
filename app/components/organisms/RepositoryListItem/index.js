/**
 * RepositoryListItem
 *
 * Lists the name and the issue count of a repository
 */

import React, {useMemo} from 'react'
import PropTypes from 'prop-types'
import {FormattedNumber} from 'react-intl'
import {useSelector} from 'react-redux'

import ListItem from 'components/molecules/ListItem'
import useRepositoryState from 'hooks/repository/entity/useState'
import {selectors as homePageSelectors} from 'modules/ui/username-form'

import IssueIcon from './IssueIcon'
import IssueLink from './IssueLink'
import RepositoryLink from './RepositoryLink'
import Wrapper from './Wrapper'

const RepositoryListItem = ({id}) => {
  const repository = useRepositoryState({id})
  const username = useSelector(homePageSelectors.selectUsername)

  /*
   * If the repository is owned by a different person than we got the data for
   * it's a fork and we should show the name of the owner
   */
  const namePrefix = useMemo(() => {
    if (repository.ownerUsername !== username) {
      return `${repository.ownerUsername}/`
    }
    return ''
  }, [id, username, repository.ownerUsername])

  const content = (
    <Wrapper>
      <RepositoryLink href={repository.url} target="_blank">
        {namePrefix + repository.name}
      </RepositoryLink>
      <IssueLink href={`${repository.url}/issues`} target="_blank">
        <IssueIcon />
        <FormattedNumber value={repository.openIssuesCount} />
      </IssueLink>
    </Wrapper>
  )

  return (
    <ListItem key={`repo-list-item-${repository.fullName}`} item={content} />
  )
}

RepositoryListItem.propTypes = {
  id: PropTypes.string.isRequired
}

export default React.memo(RepositoryListItem)
