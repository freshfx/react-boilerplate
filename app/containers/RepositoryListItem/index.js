/**
 * RepositoryListItem
 *
 * Lists the name and the issue count of a repository
 */

import React, {useMemo} from 'react'
import PropTypes from 'prop-types'
import {FormattedNumber} from 'react-intl'

import ListItem from 'components/ListItem'

import IssueIcon from './IssueIcon'
import IssueLink from './IssueLink'
import RepositoryLink from './RepositoryLink'
import Wrapper from './Wrapper'
import useRepository from './useRepository'

const RepositoryListItem = ({id}) => {
  const [state] = useRepository(id)

  /*
   * If the repository is owned by a different person than we got the data for
   * it's a fork and we should show the name of the owner
   */
  const namePrefix = useMemo(() => {
    if (state.ownerUsername !== state.username) {
      return `${state.ownerUsername}/`
    }
    return ''
  }, [id])

  const content = (
    <Wrapper>
      <RepositoryLink href={state.url} target="_blank">
        {namePrefix + state.name}
      </RepositoryLink>
      <IssueLink href={`${state.url}/issues`} target="_blank">
        <IssueIcon />
        <FormattedNumber value={state.openIssuesCount} />
      </IssueLink>
    </Wrapper>
  )

  return <ListItem key={`repo-list-item-${state.fullName}`} item={content} />
}

RepositoryListItem.propTypes = {
  id: PropTypes.string.isRequired
}

export default React.memo(RepositoryListItem)
