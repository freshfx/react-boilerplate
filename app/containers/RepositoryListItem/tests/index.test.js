import React from 'react'
import {render} from '@testing-library/react'
import {IntlProvider} from 'react-intl'

import RepositoryListItem from '../index'
import useRepository from '../useRepository'

const mockedState = {
  fullName: 'react-boilerplate/react-boilerplate',
  name: 'react-boilerplate',
  openIssuesCount: 20,
  ownerUsername: 'mxstbr',
  url: 'https://github.com/react-boilerplate/react-boilerplate',
  username: 'mxstbr'
}
jest.mock('../useRepository')
jest.mock('components/ListItem', () => ({item}) => item)

const id = 'repository-id-1'
const renderComponent = (props = {}) => render(
  <IntlProvider locale="en">
    <RepositoryListItem id={id} {...props} />
  </IntlProvider>
)

describe('RepositoryListItem', () => {
  beforeAll(() => {
    useRepository.mockImplementation(() => [mockedState])
  })

  it('should match snapshot', () => {
    const {container} = renderComponent()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should add nameprefix if user is not currentUser', () => {
    const username = 'ffx'
    useRepository.mockImplementationOnce(() => [{...mockedState, username}])
    const {getByText} = renderComponent()
    expect(getByText(`${mockedState.ownerUsername}/${mockedState.name}`))
  })
})
