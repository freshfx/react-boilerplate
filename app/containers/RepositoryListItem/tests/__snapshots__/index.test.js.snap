// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryListItem should match snapshot 1`] = `
.emotion-6 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: space-between;
  -webkit-box-align: space-between;
  -ms-flex-align: space-between;
  align-items: space-between;
}

.emotion-0 {
  color: #41addd;
  height: 100%;
  color: black;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-0:hover {
  color: #6cc0e5;
}

.emotion-4 {
  color: #41addd;
  height: 100%;
  color: black;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.emotion-4:hover {
  color: #6cc0e5;
}

.emotion-2 {
  fill: #ccc;
  margin-right: 0.25em;
}

<div
  class="emotion-6 emotion-7"
>
  <a
    class="emotion-0 emotion-1"
    href="https://github.com/react-boilerplate/react-boilerplate"
    target="_blank"
  >
    react-boilerplate
  </a>
  <a
    class="emotion-4 emotion-5"
    href="https://github.com/react-boilerplate/react-boilerplate/issues"
    target="_blank"
  >
    <svg
      class=" emotion-2 emotion-3"
      height="1em"
      width="0.875em"
    >
      <path
        d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7S10.14 13.7 7 13.7 1.3 11.14 1.3 8s2.56-5.7 5.7-5.7m0-1.3C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7S10.86 1 7 1z m1 3H6v5h2V4z m0 6H6v2h2V10z"
      />
    </svg>
    20
  </a>
</div>
`;
